### DISCUSSION TITLE
Unable to cast object of type 'Avalonia.Media.GeometryDrawing' to type 'Avalonia.IStyledElement'.'

### DISCUSSION THREAD

**Discussion by elveejay1** [AvaloniaUI/Avalonia#9145] [Q&A] [answered]
In a button of my own design, I want to be able to show some SVG code as an icon.

My style looks like this:
```

	<Style Selector="controls|TouchButton">
		<Setter Property="Template">
			<ControlTemplate TargetType="{x:Type controls:TouchButton}">
				<Button Height="{TemplateBinding Height}" 
						Width="{TemplateBinding Width}" 
						Background="{TemplateBinding Background}" 
						CornerRadius="0" 
						HorizontalContentAlignment="Center" 
						VerticalContentAlignment="Center"
						Command="{TemplateBinding Command}">
					<StackPanel Orientation="Horizontal" Margin="10,5,10,5">
						<StackPanel.Resources>
							<DrawingImage x:Key="PART_IconGeometryDrawing">
								<DrawingImage.Drawing>
									<DrawingGroup>
										<GeometryDrawing Brush="{TemplateBinding Foreground}" Geometry="{TemplateBinding Icon}"/>
									</DrawingGroup>
								</DrawingImage.Drawing>
							</DrawingImage>
						</StackPanel.Resources>
						<Image Height="{TemplateBinding FontSize}" Source="{StaticResource PART_IconGeometryDrawing}"/>
						<TextBlock Text="{TemplateBinding Caption}" 
								   Foreground="{TemplateBinding Foreground}" 
								   FontSize="{TemplateBinding FontSize}"/>
					</StackPanel>
				</Button>
			</ControlTemplate>
		</Setter>
	</Style>

```

My property:
```

        /// <summary>
        /// Icon Geometry
        /// </summary>
        public Geometry Icon
        {
            get { return (Geometry)GetValue(IconProperty); }
            set { SetValue(IconProperty, value); }
        }
        public static readonly StyledProperty<Geometry> IconProperty =
           AvaloniaProperty.Register<TouchButton, Geometry>(nameof(Icon));

```

But for some reason I get the exception 
Unable to cast object of type 'Avalonia.Media.GeometryDrawing' to type 'Avalonia.IStyledElement'.'

What am I doing wrong? The other properties bind nicely, though.

I also tried using 
{Binding Icon, RelativeSource={RelativeSource TemplatedParent}}
as mentioned in this link (I know it is WPF, but I hoped it worked):
https://stackoverflow.com/questions/9591228/is-template-binding-to-a-geometry-possible

**Comment by elveejay1** [comment, marked as answer]
I think I found a way - thought I'd share my solution if others have the same issue.

I made this interface:
```

    public interface IGeometryDrawingControl
    {
        public Geometry Icon { get; set; }

        public IBrush Foreground { get; set; }
    }

```

Not really needed but allows me to re-use this converter:

```

    public class ImageSourceConverter : IValueConverter
    {
        public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
        {
            IGeometryDrawingControl gdc = value as IGeometryDrawingControl;

            if (gdc == null) return null;

            GeometryDrawing g = new GeometryDrawing();
            g.Geometry = gdc.Icon;
            g.Brush = gdc.Foreground;

            return new DrawingImage(g);
        }

        public object? ConvertBack(object? value, Type targetType, object? parameter, CultureInfo culture)
        {
            throw new NotImplementedException();
        }
    }

```

And then change my Image to look like this:

```
<Image Height="{TemplateBinding FontSize}" Margin="0,0,4,0" Source="{TemplateBinding Converter={StaticResource ImageSourceConverter}}"/>

```

**Comment by rabbitism** [comment]
Just wondering, why not set Image source directly like this?

```xml
<Image Height="{TemplateBinding FontSize}">
    <Image.Source>
        <DrawingImage>
            <DrawingImage.Drawing>
                <DrawingGroup>
                    <GeometryDrawing Brush="{TemplateBinding Foreground}" Geometry="{TemplateBinding Icon}"/>
                </DrawingGroup>
            </DrawingImage.Drawing>
        </DrawingImage>
    </Image.Source>
</Image>
```

**Reply by elveejay1** [reply]
That gave me the exception mentioned above.

**Reply by rabbitism** [reply]
Yeah, it's a bit off topic, let me reply the answer separately.

**Comment by rabbitism** [comment]
`GeometryDrawing` inherits `AvaloniaObject` directly, so it doesn't implement `IStyleElement`. 
When using `TemplateBinding` syntax, the TemplateBinding class will cast target to `IStyleElement`, see
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Markup/Avalonia.Markup/Data/TemplateBinding.cs#L48 
So the failure not only happens to this Geometry, but also Brush. 
You need to change them all to `{Binding RelativeResource={RelativeResource TemplateParent}}` syntax. 

I think this `TemplateBinding` limitation should be mentioned in documentation. You can raise an issue there or I can help if you don't want to.

**Reply by elveejay1** [reply]
I tried {Binding Icon, RelativeSource={RelativeSource TemplatedParent}}, which gave me the same error. I have no idea how to raise this issue, but if it is worthy of raising, please do. Apparently that solution worked for someone who had a similar problem in WPF...so it would make sense I guess for it to work here also.

**Reply by rabbitism** [reply]
> I tried {Binding Icon, RelativeSource={RelativeSource TemplatedParent}}, which gave me the same error. I have no idea how to raise this issue, but if it is worthy of raising, please do. Apparently that solution worked for someone who had a similar problem in WPF...so it would make sense I guess for it to work here also.

Because you brush is still using `TemplateBinding`, you need to change that as well.
